Updated Python plotting script for UFS SRW app

BenjaminBlake-NOAA opened the pull request:
## DESCRIPTION OF CHANGES: 
This is an updated version of the Python plotting script for the UFS SRW app.  The script now uses the argparse library, reads in an env.yaml file (currently blank), defines the location of the input grib2 file with placeholders (they will need to be read in somehow), and uses cartopy instead of basemap.  I have never used cartopy before and I think the plots definitely need to be improved but this was a first attempt.  If anyone knows more about cartopy than I do (which is very little!) please feel free to update/adjust accordingly.

## TESTS CONDUCTED: 
None required.

dmwright526 commented:
Hi @BenjaminBlake-NOAA, I am working on trying this script out on Jet while getting my python environment setup there. Is the goal of the .yaml file to define the variables EXPT_BASEDIR and EXPT_SUBDIR? 

I have a little bit of experience using cartopy so I will see if there is anything I can add...

BenjaminBlake-NOAA commented:
Hi @dmwright526 thanks for testing out the script and yes feel free to add anything you think may be helpful!  I've mostly worked with basemap up to this point.  I think the eventual goal of the .yaml is to help define the Python virtual environment, EXPT_BASEDIR and EXPT_SUBDIR are placeholders for variables which are defined in the workflow.  @JeffBeck-NOAA or @gsketefian can probably speak more to how those variables will get passed in.

dmwright526 commented:
@BenjaminBlake-NOAA sounds good. Right now I have it working with the .yaml file able to pass in the EXPT_BASEDIR and EXPT_SUBDIR variables.
As for cartopy, I am working on specifying the path for the shapefiles download so the script works across all systems. I got the addition of state boundaries into the maps through uploading my own copy of the shapefiles and sourcing them, now it is just a matter of getting them to download properly on a new build of cartopy.  I should hopefully have that fixed in the next few days.

JeffBeck-NOAA commented:
Thanks, @dmwright526!  Once you have something finalized, please feel free to push a commit to the feature/python branch in Ben's fork!  After Ben's accepted it, I'll be happy to test on Hera too.

dmwright526 commented:
@JeffBeck-NOAA a few changes have been made to the plotting script with instructions on how to run it in plot_allvars.py.

JeffBeck-NOAA commented:
Thanks, @dmwright526!  I'll check it out and test things on Hera.

JeffBeck-NOAA commented:
@BenjaminBlake-NOAA @dmwright526 I'm having problems loading the correct anaconda on Hera to get the script to run.  I tried:

module use /contrib/modulefiles
module load anaconda/latest

However, this fails. as /contrib/modulefiles is unreadable and anaconda/latest no longer exists.  @BenjaminBlake-NOAA Do you know what we should use instead?  Also, did we decide if/how we wanted to document the necessary module load commands for the Tier 1 platforms to run the script?

BenjaminBlake-NOAA commented:
Hi @JeffBeck-NOAA  a couple weeks ago they changed the modulefiles in contrib on Hera.  I think it'll work if you do this instead:
module use /contrib/anaconda/modulefiles
module load anaconda/latest

I do not think we've decided on how to add that documentation just yet.

JeffBeck-NOAA commented:
@BenjaminBlake-NOAA Thanks!  That worked.  The PROJ_LIB environment variable must also be set on Hera (I don't know about Jet) as /contrib/anaconda/anaconda3/latest/share/proj/.  

I was able to get plots of fields for a CONUS domain!  I noticed that as before, the script won't run with rotated lat/lon , but that's fine since it won't be a supported pre-defined domain for the release.  We're only supporting 3-km/13-km/25-km CONUS.

@BenjaminBlake-NOAA @dmwright526 A couple things I'd like to submit in a commit, if everyone's OK with it:

1) Modify the file names to remove "compare"
2) Modify the title from FV3SAR to FV3-LAM, since that's the (brand new!) acronym we're going to use moving forward
3) Change the default path to: /path/to/exp/dir/'+str(ymdh)+'/postprd/RRFS.t'+cyc+'z.bgdawp'+fhour+'.tm00'

dmwright526 commented:
@JeffBeck-NOAA Those updates sound find to me. For the default path, are the last 2 characters always '00' or should those be cycle hours as well?

BenjaminBlake-NOAA commented:
@JeffBeck-NOAA  I forgot to respond earlier today but yes that sounds good to me too!  @dmwright526  I think it should always be tm00 at the end but Jeff can confirm.

dmwright526 commented:
@JeffBeck-NOAA @BenjaminBlake-NOAA I just went and updated what was suggested to try to get this PR wrapped up.

On a side note, it looks like from Chan-Hoo Jeon's FV3 document (section 10.2.1, page 144), the shape files for Cartopy are on Hera, Orion, and WCOSS. So if you point the CARTOPY_DIR variable to those locations and add in line ax.add_feature... (line 424), you should be able to plot the 50m resolution state boundaries.

JeffBeck-NOAA commented:
@dmwright526 @BenjaminBlake-NOAA David, I'm sorry for not getting to those edits sooner, things have been crazy with planning the UFS Users' Workshop and the release.  Did you push a commit with the changes?  I haven't seen it yet.   If you haven't started, I can still try to squeeze it in this weekend.

dmwright526 commented:
Sorry @JeffBeck-NOAA , my push evidently didn't go through... I just did it again and it now shows up.

christinaholtNOAA commented on ush/Python/env.yaml:
```diff
@@ -0,0 +1,6 @@
+# This is a yaml file to define the Python virtual environment
```
An env.yaml file should be reserved for preparing a conda environment. If you need these environment variables, please consider using a different naming convention for this file, or using argparse to ingest all your information.

More information about conda's environment.yml can be found here: https://www.google.com/search?q=conda+environment+from+yml&oq=conda+environment+from+yml&aqs=chrome..69i57j0l7.11670j1j7&sourceid=chrome&ie=UTF-8

You'd want to put conda environment installation information in this file, including only the external python packages that are necessary. For example, this script requires matplotlib, cartopy, pygrib, pyyaml, etc. A linter (like pylint) could be useful in ensuring that you have only the imports you need.

BenjaminBlake-NOAA commented:
@christinaholtNOAA  Thanks for your suggestions and we could pass in those environment variables as arguments instead.  @JeffBeck-NOAA  is that the approach we should take?  @dmwright526 with regards to the cartopy shapefiles, I can update the script to point to those.

JeffBeck-NOAA commented:
@BenjaminBlake-NOAA Yes, it sounds like we should pass in these arguments with "argparse."  Is that something you are familiar with and could implement?

BenjaminBlake-NOAA commented:
@JeffBeck-NOAA  I think so and I could probably do that tomorrow morning.  Once I have an updated version ready I'll commit that.

JeffBeck-NOAA commented:
Thanks, @BenjaminBlake-NOAA!

BenjaminBlake-NOAA commented:
I just committed a new version of the script which removes the variables from env.yaml and defines them as command line arguments.  I also added code for using the cartopy shapefiles which makes the plots look much nicer.  To use them on Hera, you can set CARTOPY_DIR = /scratch2/NCEPDEV/fv3-cam/Chan-hoo.Jeon/tools/NaturalEarth.  @dmwright526  I added your name as an author at the top, is University of Michigan your affiliation?

dmwright526 commented:
Thanks, @BenjaminBlake-NOAA . Yes, University of Michigan is my affiliation. I will give the script a try on Jet tomorrow morning when it comes back online.

JeffBeck-NOAA commented:
@BenjaminBlake-NOAA @dmwright526 Do we need the env.yaml file now that it's empty?  It's not clear to me what it's use would be now.

BenjaminBlake-NOAA commented:
@JeffBeck-NOAA  It is currently empty, but I'm not sure if it will be needed to prepare the conda environment (as @christinaholtNOAA described above) so I left it there for now.

JeffBeck-NOAA commented:
@dmwright526 @BenjaminBlake-NOAA I just test this on Hera and it looks great.  Thanks for getting the changes applied and adding the cartopy shape files!  A couple tweaks that still need to be made:

1) Change 'tm00' to 'tm'+cyc at the end of the file path for the grib2 files
2) Remove the "compare" string in the plot file names.

BenjaminBlake-NOAA commented:
@JeffBeck-NOAA  Thanks for testing the script!

1. I saw that someone had made that change but I was confused about it, so I removed it.  tmmark generally refers to time minus HH hours from the advertised cycle time, so without data assimilation we should always be at tm00.  If the grib2 files are named 'tm'+cyc then I propose we consider changing that to avoid confusion with tmmark.  What do you think?
2. I can certainly do this after we decide about how to handle #1.

JeffBeck-NOAA commented:
Hey @BenjaminBlake-NOAA,

I took a look at our ex-script for post and saw the following:

yyyymmdd=${CDATE:0:8}
hh=${CDATE:8:2}
cyc=$hh
tmmark="tm$hh"

Therefore, we're currently defining "tmmark" as "cyc".  We should definitely get it in line with what is done in operations.  How is "tmmark" defined in the ex-script for HRRR operations on WCOSS?  The HRRRX doesn't use the "tmmark" convention, so it's not in the post script on Jet.

BenjaminBlake-NOAA commented:
Hi @JeffBeck-NOAA , the operational HRRR does not use tmmark either.  The NAM uses tmmark="tm??" where the ?? is defined as the number of hours offset from the cycle start time.  For example, offset=-`echo $tmmark | cut -c 3-4`.  Since we will always have an offset of 0 hours for cold start runs w/ no DA, it should be tm00.  So I suggest we either do something similar as the NAM, or we remove tmmark entirely.  I'm not opposed to removing tmmark and I think it would be less confusing without it there.  Maybe we could change RRFS.t${cyc}z.bgdawp${fhr}.${tmmark} to RRFS.t${cyc}z.bgdawp${fhr}.grib2 to make it more clear that it's a grib2 file.  @RatkoVasic-NOAA @JacobCarley-NOAA  what do you both think?

JeffBeck-NOAA commented:
@BenjaminBlake-NOAA I agree that to limit confusion, we may want to remove the tmmark value and add .grib2 to make it clear what type of file is being output by UPP.  I'm in favor of that change.  Let's see what @RatkoVasic-NOAA and @JacobCarley-NOAA think.

JeffBeck-NOAA commented:
@BenjaminBlake-NOAA Based on our Slack discussion, what do you think about using **RRFS.t${cyc}z.bgdawp${fhr}.tm00.grib2** for now?  We can reassess when we get DA/partial cycling in the workflow.

BenjaminBlake-NOAA commented:
@JeffBeck-NOAA  RRFS.t${cyc}z.bgdawp${fhr}.tm00.grib2 sounds good to me.  I just pushed that change to my branch.  I also removed the compare strings from the plot file names.

JacobCarley-NOAA commented:
Quick comment - I think the "RRFS" in the filename should be lowercase as that's convention in operations.

BenjaminBlake-NOAA commented:
@JacobCarley-NOAA  That is true, good point.  I'll make that change now.

BenjaminBlake-NOAA commented:
@JeffBeck-NOAA  Sounds good to me!  I'll do that now.
